src/provider.cpp

```diff
@@ -54,6 +54,47 @@ wsrep::provider* wsrep::provider::make_provider(
     return 0;
 }
 
+std::string
+wsrep::provider::status_str(enum wsrep::provider::status const val)
+{
+    std::ostringstream os;
+
+    switch(val)
+    {
+    case success:
+        os << "Success"; break;
+    case error_warning:
+        os << "Warning"; break;
+    case error_transaction_missing:
+        os << "Transaction not registered with provider"; break;
+    case error_certification_failed:
+        os << "Certification failed"; break;
+    case error_bf_abort:
+        os << "Transaction was BF aborted"; break;
+    case error_size_exceeded:
+        os << "Transaction size exceeded"; break;
+    case error_connection_failed:
+        os << "Not connected to Primary Component"; break;
+    case error_provider_failed:
+        os << "Provider in bad state, needs to be reinitialized."; break;
+    case error_fatal:
+        os << "Fatal error, must abort."; break;
+    case error_not_implemented:
+        os << "Function not implemented"; break;
+    case error_not_allowed:
+        os << "Operation not allowed"; break;
+    case error_unknown:
+        os << "Unknown error"; break;
+    }
+
+    /* Just in case we have a negative error code treat it as a system code */
```

Comment:
Do we really have API calls which may return both `enum wsrep_status` and negative error code? Doesn't that defeat the purpose of having enum return type? At least I would be more verbose here about violating type safety.

Reply:
I can't vouch for every call. I;m quite sure at least some calls that return `int`may return negative error code. So precisely because this may be a possibility and it is wrong, if find it quite unhelpful to hide it behind `unknown` error. Very often we won't have a second chance to investigate it.

Reply:
What sort of verbosity do you have in mind?

Reply:
> I can't vouch for every call. I;m quite sure at least some calls that return `int`may return negative error code. So precisely because this may be a possibility and it is wrong, if find it quite unhelpful to hide it behind `unknown` error. Very often we won't have a second chance to investigate it.

`int` return values should not be converted to `wsrep::provider::status` codes. Meaning that `map_return_value()` below should be used only to functions which return `enum wsrep_status`. Return values from functions which return `int` should be handled separately, and I believe the convention there is to return zero in case of success or negative error code. Therefore we should never see here other than enum values.

Moreover, the signedness of the underlying enum value may vary depending on the compiler implementation, so the check below is not correct anyway.

I would change this to default case in switch above which would print `unknown error code` and raw value into `os', combined with warning log message.